[14.0] Shopfloor imp zone picking and cluster picking

TDu opened the pull request:
ref.: rau-44

sebalix committed:
shopfloor: imp zp scan source package message

TDu committed:
shopfloor: imp zp scan source location

Fix the error raised when there are multiple lots in location.
In case of error, reduce the list to the new location scanned
and displayed lines are filtered accordingly.

TDu committed:
shopfloor: imp zp scan source product

Ff multiple move lines are found on several locations, raise an error
and filter the list with the matched moves.

If multiple move lines are found for different lots, raise an error
and filter the list with the matched moves.

TDu committed:
shopfloor: imp zp scan source lot

If multiple move lines are found on several locations, display a message
and filter the list with the matched moves.
If multiple move lines are found for different products, raise an error "Several products found, please scan first a product"

TDu committed:
shopfloor: imp cluster picking  _scan_line_by_package

TDu committed:
shopfloor: fix cluster picking _scan_line_by_product

Check the content of the location not the move lines of the picking.

TDu committed:
shopfloor: fix cluster picking _scan_line_location

Fix the error raised if there are multiple packs, products or lot in the location.
Not based on move lines but based on actual stock at location.

TDu committed:
shopfloor: fix cluster picking _scan_line_lot

Take into account the content of the location not move on the picking
when checking for lot in multiple packages.

jbaudoux commented on shopfloor/actions/message.py:
```diff
@@ -256,6 +276,12 @@ def several_products_in_location(self, location):
+    def several_products_in_package(self, package):
+        return {
+            "message_type": "error",
+            "body": _("Several products found in %s, aborting." % package.name),
```
```suggestion
            "body": _("Several products found in %s, please scan the product." % package.name),
```

jbaudoux commented on shopfloor/actions/message.py:
```diff
@@ -248,6 +248,26 @@ def several_lots_in_location(self, location):
+    def several_lots_in_package(self, package):
+        return {
+            "message_type": "error",
+            "body": _("Several lots found in %s, aborting." % package.name),
```
```suggestion
            "body": _("Several lots found in %s, please scan the lot." % package.name),
```

jbaudoux commented on shopfloor/actions/message.py:
```diff
@@ -248,6 +248,26 @@ def several_lots_in_location(self, location):
+        return {
+            "message_type": "warning",
+            "body": _(
+                "Several moves found on different locations, please scan first a location."
```
```suggestion
                "Several moves found on different locations, please scan first the location."
```

jbaudoux commented on shopfloor/actions/message.py:
```diff
@@ -248,6 +248,26 @@ def several_lots_in_location(self, location):
+    def several_move_with_different_lot(self):
+        return {
+            "message_type": "warning",
+            "body": _("Several moves found for different lots, please scan a lot."),
```
```suggestion
            "body": _("Several moves found for different lots, please scan the lot."),
```

jbaudoux commented on shopfloor/services/cluster_picking.py:
```diff
@@ -509,13 +527,16 @@ def _scan_line_by_lot(self, picking, move_line, lot):
+        location_quants = move_line.location_id.quant_ids.filtered(
+            lambda quant: quant.quantity > 0 and quant.lot_id == lot
+        )
+        packages = location_quants.mapped("package_id")
```
No need for mapped in v14

jbaudoux commented on shopfloor/services/cluster_picking.py:
```diff
@@ -530,21 +551,17 @@ def _scan_line_by_location(self, picking, move_line, location):
+        location_quants = move_line.location_id.quant_ids.filtered(
+            lambda quant: quant.quantity > 0
+        )
+        lots = location_quants.mapped("lot_id")
```
No need for mapped in v14

jbaudoux commented on shopfloor/services/cluster_picking.py:
```diff
@@ -530,21 +551,17 @@ def _scan_line_by_location(self, picking, move_line, location):
-        packages = pending_lines.mapped("package_id")
-        products = pending_lines.mapped("product_id")
-
+        packages = location_quants.mapped("package_id")
```
No need for mapped in v14

jbaudoux commented on shopfloor/services/cluster_picking.py:
```diff
@@ -530,21 +551,17 @@ def _scan_line_by_location(self, picking, move_line, location):
-        products = pending_lines.mapped("product_id")
-
+        packages = location_quants.mapped("package_id")
+        products = location_quants.mapped("product_id")
```
No need for mapped in v14

jbaudoux commented on shopfloor/services/zone_picking.py:
```diff
@@ -495,15 +511,18 @@ def _scan_source_location(self, barcode, confirmation=False):
+    def _scan_source_package(self, barcode, confirmation=False, product_id=False):
         """Search a package and find available lines for it.
 
         Fist search for lines that have the specific package.
```
```suggestion
        First search for lines that have the specific package.
```

jbaudoux commented on shopfloor/services/zone_picking.py:
```diff
@@ -555,32 +581,44 @@ def _scan_source_product(self, barcode, confirmation=False):
             return response, message
         move_lines = self._find_location_move_lines(product=product)
-        if move_lines:
+        if len(move_lines.mapped("location_id")) > 1:
```
No need for mapped in v14

jbaudoux commented on shopfloor/services/zone_picking.py:
```diff
@@ -555,32 +581,44 @@ def _scan_source_product(self, barcode, confirmation=False):
-            if move_lines:
+            if not move_lines:
+                continue
+            if len(move_lines.mapped("location_id")) > 1:
```
No need for mapped in v14

jbaudoux commented on shopfloor/tests/test_cluster_picking_scan_line.py:
```diff
@@ -72,10 +102,20 @@ def test_scan_line_serial_ok(self):
-        """Scan a product tracked by lot, must scan the lot"""
+        """Scan a product tracked by lot, must scan the lot.
+
+        If for the same product there is multiple lot in the location.
+        The user need to scan the lot.
```
```suggestion
        If for the same product there is multiple lots in the location,
        the user must scan the lot.
```

sebalix replied:
But if the only product available in the operation processed is in only one lot among reserved move lines, and in a second lot in another operation for the same location (but we don't care about), why should we block the scan? :thinking:

jbaudoux commented on shopfloor/tests/test_cluster_picking_scan_line.py:
```diff
@@ -260,7 +312,7 @@ def test_scan_line_location_error_several_products(self):
         location = line.location_id
         # add a second product in the location
-        self._update_qty_in_location(location, self.product_b, 10)
+        # self._update_qty_in_location(location, self.product_b, 10)
```
on purpose?

jbaudoux commented on shopfloor/tests/test_cluster_picking_scan_line.py:
```diff
@@ -288,7 +340,9 @@ def test_scan_line_location_error_several_products(self):
         line.qty_done = line.product_uom_qty
-        self._scan_line_ok(new_move.move_line_ids[0], line.location_id.barcode)
+        line._action_done()
+        # FIXME the previous action_done does not work !
```
FIXME

sebalix replied:
Same remark, we should not call `_action_done` on a line, this one is called by the `stock.move._action_done()`, itself called by `stock.picking._action_done()` (which is the only method we should deal with when it comes to validate a move).

jbaudoux reviewed (commented):
Thanks for the improvements. A few minor remarks and a test you disabled with a FIXME

sebalix commented on shopfloor/tests/test_cluster_picking_scan_line.py:
```diff
@@ -249,6 +300,7 @@ def test_scan_line_location_error_several_package(self):
         # Although, if one line is already processed,
         # we can work automatically on the next one
         line.qty_done = line.product_uom_qty
+        line._action_done()
```
Calling this method at the `stock.move.line` level seems wrong, what was the intent?

sebalix commented on shopfloor/tests/test_cluster_picking_scan_line.py:
```diff
@@ -326,6 +380,7 @@ def test_scan_line_location_error_several_lots(self):
         # Although, if one line is already processed,
         # we can work automatically on the next one
         line.qty_done = line.product_uom_qty
+        line._action_done()
```
`/!\`

TDu committed:
shopfloor: fix test multiple lot,product,pack on location

The tests were like they were because the code was testing the existance
of a lot, pack or product in a location based on existing pick/moves.
But now the code tests the location directely for such information.

TDu commented:
@sebalix and @jbaudoux I have done the requested change and refactor the tests that caused problems.

jbaudoux reviewed (approved):
Thanks. LGTM

TDu commented:
I squashed the fixups.

sebalix commented on shopfloor/tests/test_zone_picking_select_line.py:
```diff
@@ -419,6 +426,34 @@ def test_scan_source_barcode_lot(self):
+            "scan_source",
+            params={"barcode": lot.name},
+        )
+        # FIX ME: need to filter lines only on lot scanned !!
+        move_lines = self.service._find_location_move_lines()
```
Is it still an issue?

TDu replied:
Yes, it seems not to be implemented, see line below as well. Activating it fail.

sebalix commented on shopfloor/tests/test_zone_picking_select_line.py:
```diff
@@ -404,6 +404,13 @@ def test_scan_source_barcode_lot(self):
         """Scan source: scanned lot has one related move line,
         next step 'set_line_destination' expected on it.
         """
+        # Product C has already one lot from test_zone_picking_base.py
+        # So lets add one more lot for that product in a different location.
+        pick = self._create_picking(lines=[(self.product_c, 10)])
+        self._fill_stock_for_moves(
+            pick.move_lines, in_lot=True, location=self.zone_sublocation1
+        )
+        pick.action_assign()
```
Why do we need to add this in this test? It didn't run well without this change? Or it is to ensure that everything is working as expected by adding another lot on a different location?

TDu replied:
It ran well without it. It was to give the test a little more context.

sebalix commented on shopfloor/services/cluster_picking.py:
```diff
@@ -490,21 +490,31 @@ def _scan_line_by_product(self, picking, move_line, product):
-        other_product_lines = picking.move_line_ids.filtered(
-            lambda l: l.product_id == product and l.location_id == move_line.location_id
+        message = None
+        location_quants = move_line.location_id.quant_ids.filtered(
+            lambda quant: quant.quantity > 0 and quant.product_id == product
         )
```
A `search` on `stock.quant` will be better I guess here, if you get thousands of quants in your location, it'll loop over all of them with `filtered`.

Also I'm wondering: we will grab quants that are not related to the current `picking`, even some reserved by other operations, is it what we want?! The packages and lots we are looking for should be those already reserved in `picking` IMO. cc @jbaudoux

sebalix replied:
This commit and the [next one](https://github.com/OCA/wms/pull/443/commits/99a7cbdc2802f141fb285d5a10815a6f2aa1f55f) are titled

```
shopfloor: fix cluster picking _scan_line_by_product

Check the content of the location not the move lines of the picking.
```
```
shopfloor: fix cluster picking _scan_line_location

Fix the error raised if there are multiple packs, products or lot in the location.
Not based on move lines but based on actual stock at location.
```
So it has been done on purpose, but I do not understand why :thinking:

jbaudoux replied:
@sebalix You are in front of a location and we want to be sure you take the right product from that location. So it must look at what is on the location, not what is in the processed picking.
I think filtered is fine, there won't be much quants on a location bin.

jbaudoux replied:
@sebalix Same for lot. If there are multiple lots on the location, you have to scan the right one. Scanning the product is only enough if there is only a single lot on the location. No matter what are the picking move lines

sebalix replied:
This implies to use location bin, with a too big source location we'll fetch too many things. Perhaps it'll never happen, I haven't the big picture, because it's like we developed this feature to fit our customers config, while it could not work well in another context. I could be wrong.

TDu replied:
Are we not filtering on the quants related to that location and move line on line 499 ?

sebalix commented:
/ocabot merge minor
